ecs-client.js

```diff
@@ -0,0 +1,102 @@
+'use strict';
+
+const { ECSClient, StopTaskCommand } = require("@aws-sdk/client-ecs");
+const axios = require('axios');
+const callWithRetries = require('./call-with-retries');
+const Exception = require("./exception");
+const exitDeferred = require('./exit-deferred');
+
+const DEFAULT_TIMEOUT = 30000;
+const DEFAULT_REASON = 'ecs-client.js: default stop reason';
+
+class EcsClient {
+
+    /**
+     * @param {Object} options
+     * @param {Object} options.aws - AWS config
+     * @param {String} options.aws.accessKeyId
+     * @param {String} options.aws.secretAccessKey
+     * @param {String} options.aws.region
+     * @param {String} options.metadataUri
+     * @param {Number} [options.timeout=30000] - used for metadata fetch request
+     * @param {Object=} options.logger - An instance of sq-logger
+     */
+    constructor(options) {
+        if(options.logger == null){
+            throw new Exception(Exception.ErrorCode.ERROR_INVALID_PARAMETER, 'An sq-logger instance must be provided.');
+        }
+        this.logger = options.logger;
+        this.metadataUri = options.metadataUri;
+        this.client = new ECSClient(options.aws);
+        this.timeout = options.timeout || DEFAULT_TIMEOUT;
+    }
+
+    /**
+     * Trigger a stop task request to ECS
+     * @param reason
+     * @returns {Promise<*>}
+     */
+    async stopTask(reason) {
+        try {
+            this.logger.info(`ecs-client.js:: Stopping task...`);
+            const metadata = await callWithRetries(3, this, this._fetchMetadata);
```

Comment:
The metadata could be fetched async on constructor and cached in order to do not waste time for the metadata on the stopping task

Reply:
This change was considered, but I opted out of it to avoid needing an init() method. Doing this async on the constructor makes handling errors when fetching the metadata difficult and there isn't much of a downside if stopTask takes a second longer to execute (the delay is negligible when compared to the time it takes for ECS to deregister the task)